### DISCUSSION TITLE
Adding Support for Renesas RH850-U2A to Bao Hypervisor

### DISCUSSION THREAD

**Discussion by noobsplzwin** [bao-project/bao-hypervisor#201] [General]
Hello Bao-project team,

Renesas released the RH850-U2A, an automotive chip that supports virtualization assets. We are interested in deploying Bao Hypervisor on this platform.

Additionally, we are keen to contribute our work back to the Bao mainline release, as personally, I've been using Bao pretty heavily in my research.

We would like to understand the procedure for adding a new MCU architecture to Bao. Specifically:

Is it feasible to add support for the Renesas RH850-U2A architecture? What steps are involved in this process? What would be a rough estimate of the time and effort required for this integration for two men?

 We would appreciate your direction if any guidelines or instructions are available for this process.

You can find more details about the RH850-U2A series here: [RH850-U2A Product Page](https://www.renesas.com/en/products/microcontrollers-microprocessors/rh850-automotive-mcus/rh850u2a-zonedomain-microcontroller-series).

Looking forward to your advice and guidance.

Best regards,
Leo

**Comment by josecm** [comment]
Hello @aiweier1231 

Very glad to hear of your interest in Bao and that you willing to make such an big contribution!

If, as you say, this architecture has virtualization support it is very likely that porting Bao to it is possible. We are in the process of porting Bao to Infineon's TC4 which seems to be a very similar architecture as it falls under the umbrella of our most recent MPU-based targets.

Porting Bao to a new architecture essentially consists in implementing all the architecture specific hooks that you will find in other architectures under the *arch* folder. This includes:

1. Writing all low-level assembly code including bootstrap for setting up processor state and C environment, context-save/restore upon hypervisor entry and exit, and maybe a couple others;

2. Implementing drivers for architecture-specific devices which are essential for virtualization, mainly the interrupt controller, and implementing the hooks for the core submodules (mainly *interrupts*);

3. Implementing the VM/vCPU architecture specific context and operations (*vm submodule*) and hypervisor initialization (*vmm submodule*);

4. Implementing exception handling and forwarding for your architecture: e.g., findind out reason for the trap, decoding instructions for emulation, etc (e.g., in Arm, the *abort.c* file);

5. Implementing the hooks for other core submodules such as *cpu* and *mem* with the specific operations and requirements for your target architecture;

6. Implementing IO protection for your target architecture, if it supports any kind of firewall/IOMPU or IOMMU.

The effort on all of this depends on how well your target architecture fits Bao's design.

We will be glad to provide you with any support you might require.
